Remote debugging

revel8n opened the pull request:
Added extended support for connecting to Dolphin and debugging remotely through other applications.
- Extended breakpoint support to keep track of last reason execution paused
- Added GDBThread class, an extended form of the GDBStub implementation that was already present, in order to expose more elements of the emulation state to remote clients as well as integrate with both Interpreter and JIT runtimes
- Added TCPGecko class, an implementation of the USBGecko protocol over TCP, integrated more directly with the emulator

Wanted to see if this was an interesting addition for others, as well as possibly discuss any improvements to the implementation to better integrate with Dolphin in either case.

In particular:
- Currently remote clients can access the full memory range, even some addresses that cause PanicAlerts (left that way so as to not compromise those checks).  Only real drawback of this is the MessageBox loop until the read it finished.  Can probably work around this to just return what it would have without the PanicAlerts in place, or simply limit the remote process to the acceptable ranges of memory and read/write larger chunks of data at once instead of byte by byte.


There are two companion projects:
- dolphindbg (https://github.com/revel8n/dolphindbg) which is a plugin for IDA Pro that allows connections to the GDB interface
- geckowii (https://github.com/revel8n/geckowii) modified version of Gecko.Net which supports connecting over TCP, and works with the TCP interface provided

i hope to add plugins for other debuggers as well as extend the plugins to work with actual hardware where applicable.

Any thoughts or other information you have would be greatly appreciated.

ZexaronS commented:
Sorry if it sounds way too optimistic, but since running dolphin with full debugging is super slow ... what about some kind of a online debugging in which several dolphin devs could hook their PCs to to help with their CPU power - this ofcourse probably isn't that, but it sounds as if it could be extended or approaching to that, no ? 

I distinctly remember seeing a video where PCs are connected to a LAN like that in some enterprise server farms.

leoetlino commented on Source/Core/Core/PowerPC/BreakPoints.h:
```diff
@@ -45,6 +45,14 @@ struct TWatch
   bool is_enabled = false;
 };
 
+enum MemCheckCondition
```
Please use an enum class here:

    enum class MemCheckCondition
    {
      None,
      Read,
      ReadWrite,
    };

leoetlino commented on Source/Core/Core/PowerPC/BreakPoints.h:
```diff
@@ -69,8 +77,26 @@ class BreakPoints
   void Clear();
   void ClearAllTemporary();
 
+  void SetBreakpointTriggered(bool b, u64 addr = -1, MemCheckCondition cond = MEMCHECK_NONE)
```
`b` should have a more descriptive name. Also, you should probably use a std::optional instead of using a special value (-1).

leoetlino commented on Source/Core/Core/PowerPC/BreakPoints.h:
```diff
@@ -69,8 +77,26 @@ class BreakPoints
+      m_breakpointAddress = addr;
+      m_breakpointCondition = cond;
+  };
+  bool GetBreakpointTriggered() { return m_breakpointTriggered; };
```
This should be a const member function.

leoetlino commented on Source/Core/Core/PowerPC/BreakPoints.h:
```diff
@@ -69,8 +77,26 @@ class BreakPoints
 
+  void SetBreakpointTriggered(bool b, u64 addr = -1, MemCheckCondition cond = MEMCHECK_NONE)
+  {
+      m_breakpointTriggered = b;
```
This should be in the implementation file (.cpp), not in the header. By keeping this in the header, you will have to rebuild everything that includes this header whenever it changes.

leoetlino commented on Source/Core/Core/PowerPC/BreakPoints.h:
```diff
@@ -69,8 +77,26 @@ class BreakPoints
+      m_breakpointCondition = cond;
+  };
+  bool GetBreakpointTriggered() { return m_breakpointTriggered; };
+  bool GetBreakpointTriggered(u64& addr, MemCheckCondition& cond)
```
Please return these by value instead, maybe in a std::tuple.

leoetlino commented on Source/Core/Core/PowerPC/BreakPoints.h:
```diff
@@ -69,8 +77,26 @@ class BreakPoints
 private:
   TBreakPoints m_breakpoints;
+
+  bool m_breakpointTriggered;
```
Our naming convention for member variables is snake case, so this should be `m_breakpoint_triggered`. 

Please read our [contributing guide](https://github.com/dolphin-emu/dolphin/tree/master/Contributing.md) for more information.

leoetlino commented on Source/Core/Common/Logging/LogManager.cpp:
```diff
@@ -94,6 +94,7 @@ LogManager::LogManager()
   m_log[LogTypes::EXPANSIONINTERFACE] = {"EXI", "Expansion Interface"};
   m_log[LogTypes::FILEMON] = {"FileMon", "File Monitor"};
   m_log[LogTypes::GDB_STUB] = {"GDB_STUB", "GDB Stub"};
+  m_log[LogTypes::GDB_THREAD] = { "GDB_THREAD", "GDB Thread" };
```
Extra spaces here

(You can run clang-format to automatically fix all style issues, apart from naming)

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1350 @@
+//  GDBThread Implementations
+// --------------------------------------------------------------------------------------
+GDBThread::GDBThread() :
+    previous_state(CPU::State::PowerDown),
```
These should be [initialised in class](http://en.cppreference.com/w/cpp/language/data_members).

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1350 @@
+
+GDBThread::~GDBThread()
+{
+    Terminate();
```
We use 2 spaces for indentation.

Again, this can be fixed automatically by running clang-format, as described in the contributing guide (which I really suggest reading).

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1350 @@
+{
+    NOTICE_LOG(GDB_THREAD, "GDB thread: Initialize");
+
+    const SConfig& _CoreParameter = SConfig::GetInstance();
```
`config`

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1350 @@
+    for (i = 0; i < 16; i++)
+        p[i] = nibble2hex(u8(v >> (60 - 4 * i)));
+}
+/*
```
Unused code should not be committed.

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1350 @@
+    addr.sun_family = AF_UNIX;
+    strcpy(addr.sun_path, socket);
+
+    gdb_init_generic(PF_LOCAL, (const sockaddr *)&addr, sizeof(addr),
```
Please use `reinterpret_cast` instead of C-style casts

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1350 @@
+        (sockaddr *)&saddr_client, &client_addrlen);
+
+    saddr_client.sin_addr.s_addr = ntohl(saddr_client.sin_addr.s_addr);
+    /*if (((saddr_client.sin_addr.s_addr >> 24) & 0xff) != 127 ||
```
Should be removed.

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1350 @@
+    socklen_t client_addrlen = sizeof(saddr_client);
+
+    gdb_init_generic(PF_INET,
+        (const sockaddr *)&saddr_server, sizeof(saddr_server),
```
`reinterpret_cast`

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1350 @@
+}
+
+void GDBThread::gdb_init_generic(int domain,
+    const sockaddr *server_addr, socklen_t server_addrlen,
```
Style issues here

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1350 @@
+    WSAStartup(MAKEWORD(2, 2), &init_data);
+#endif
+
+    tmpsock = (int)socket(domain, SOCK_STREAM, 0);
```
`socket(2)` already returns an `int`, so the cast is useless here.

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1350 @@
+        ERROR_LOG(GDB_THREAD, "Failed to create gdb socket");
+
+    on = 1;
+    if (setsockopt(tmpsock, SOL_SOCKET, SO_REUSEADDR, (const char*)&on, sizeof on) < 0)
```
`reinterpret_cast<const char*>(&on)`

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1350 @@
+
+int GDBThread::gdb_data_available()
+{
+    struct timeval t;
```
`struct` is unneeded here.

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,134 @@
+    void ExecuteTaskInThread();
+
+protected:
+    bool is_running;
```
Naming: `m_is_running`, and so on.

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,134 @@
+    void OnResume();
+    void UpdateState(CPU::State current_state);
+
+    void ExecuteTaskInThread();
```
Since this is the function that runs on the separate thread, I think this name doesn't really fit.

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,134 @@
+    void gdb_signal(u32 s, u64 addr = -1, MemCheckCondition cond = MEMCHECK_NONE);
+
+protected:
+    bool connected;
```
Naming

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,134 @@
+
+    CPU::State previous_state;
+
+protected:
```
This is unnecessary, as there is already a `protected` above

leoetlino commented on Source/Core/Core/PowerPC/MMU.cpp:
```diff
@@ -562,6 +562,30 @@ void Write_F64(const double var, const u32 address)
 
+std::string Read(u32 address, size_t size)
+{
+  std::string s;
```
```
std::string s(size, '\0');
Memory::CopyFromEmu(s.data(), address, size);
```

leoetlino commented on Source/Core/Core/PowerPC/MMU.cpp:
```diff
@@ -562,6 +562,30 @@ void Write_F64(const double var, const u32 address)
+  return s;
+}
+
+void Write(u32 address, u8* data, size_t size)
```
Duplicate of `Memory::CopyToEmu`.

leoetlino commented on Source/Core/Core/PowerPC/MMU.cpp:
```diff
@@ -562,6 +562,30 @@ void Write_F64(const double var, const u32 address)
   Write_U64(integral, address);
 }
 
+std::string Read(u32 address, size_t size)
```
Duplicate of `Memory::GetString`

leoetlino commented on Source/Core/Core/PowerPC/MMU.cpp:
```diff
@@ -618,6 +642,36 @@ std::string HostGetString(u32 address, size_t size)
   return s;
 }
 
+std::string HostRead(u32 address, size_t size)
```
Duplicate of `HostGetString` (which is in this same file!)

leoetlino commented on Source/Core/Common/Logging/LogManager.cpp:
```diff
@@ -121,6 +121,7 @@ LogManager::LogManager()
   m_log[LogTypes::POWERPC] = {"PowerPC", "IBM CPU"};
   m_log[LogTypes::SERIALINTERFACE] = {"SI", "Serial Interface (SI)"};
   m_log[LogTypes::SP1] = {"SP1", "Serial Port 1"};
+  m_log[LogTypes::TCPGECKO] = { "TCPGECKO", "TCPGecko" };
```
Style

leoetlino commented on Source/Core/Core/ConfigManager.cpp:
```diff
@@ -146,6 +146,7 @@ void SConfig::SaveGeneralSettings(IniFile& ini)
   general->Set("GDBSocket", gdb_socket);
 #endif
   general->Set("GDBPort", iGDBPort);
+  general->Set("TCPGeckoPort", iTCPGeckoPort);
```
We're currently migrating to a new config system. Please add any new options to it (MainSettings.h/MainSettings.cpp) instead of adding more things to ConfigManager.

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.cpp:
```diff
@@ -0,0 +1,788 @@
+// Copyright 2016 Dolphin Emulator Project
```
2018

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.cpp:
```diff
@@ -0,0 +1,788 @@
+#include "Core/PowerPC/PowerPC.h"
+#include "Core/PowerPC/Interpreter/Interpreter_FPUtils.h"
+
+const u8 TCPGecko::GCACK = 0xAA;
```
`constexpr`

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.cpp:
```diff
@@ -0,0 +1,788 @@
+const u8 TCPGecko::GCACK = 0xAA;
+const u8 TCPGecko::GCNewVer = 0x80;
+TCPGecko::TCPGecko()
+    : client_running(false)
```
`client_running` should be in-class initialised

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.cpp:
```diff
@@ -0,0 +1,788 @@
+
+TCPGecko::~TCPGecko()
+{
+    Terminate();
```
Indentation

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.cpp:
```diff
@@ -0,0 +1,788 @@
+
+void TCPGecko::Terminate()
+{
+    if (clientThread.joinable())
```
m_client_thread

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.cpp:
```diff
@@ -0,0 +1,788 @@
+
+bool TCPGecko::Initialize()
+{
+    const SConfig& _CoreParameter = SConfig::GetInstance();
```
config

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.cpp:
```diff
@@ -0,0 +1,788 @@
+        return false;
+
+    if (!connectionThread.joinable())
+        connectionThread = std::thread(&TCPGecko::GeckoConnectionWaiter, this);
```
m_connection_thread

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.cpp:
```diff
@@ -0,0 +1,788 @@
+
+        if (IsClientValid())
+        {
+            // what's an ideal buffer size?
```
This should be removed.

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.cpp:
```diff
@@ -0,0 +1,788 @@
+        if (IsClientValid())
+        {
+            // what's an ideal buffer size?
+            char data[128];
```
For binary data, we typically use `u8` (unsigned char) instead of `char`.

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.cpp:
```diff
@@ -0,0 +1,788 @@
+                recv_fifo.insert(recv_fifo.end(), &data[0], &data[got]);
+            }
+
+            std::vector<char> packet;
```
`std::vector<u8>`

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.cpp:
```diff
@@ -0,0 +1,788 @@
+            }
+        }	// unlock transfer
+
+        if (did_nothing)
```
Is this necessary?

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.cpp:
```diff
@@ -0,0 +1,788 @@
+template<typename T>
+bool TCPGecko::read_data(T& data)
+{
+    return read_data((u8*)&data, sizeof(T));
```
reinterpret_cast

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.cpp:
```diff
@@ -0,0 +1,788 @@
+}
+
+template<typename T>
+bool TCPGecko::read_data(T& data)
```
Naming: `ReadData`

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.h:
```diff
@@ -0,0 +1,117 @@
+// Copyright 2016 Dolphin Emulator Project
```
2018

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.h:
```diff
@@ -0,0 +1,117 @@
+
+    std::queue<std::unique_ptr<sf::TcpSocket>> waiting_socks;
+
+    template<typename T>
```
I'm not sure this works.

leoetlino commented on Source/Core/Core/PowerPC/TCPGecko.h:
```diff
@@ -0,0 +1,117 @@
+        Unknown
+    };
+
+    static const u32 packetsize = 0xF800;
```
These should be in the .cpp file, as there's no reason for exposing those in the header.

leoetlino commented on Source/Core/Core/Host.h:
```diff
@@ -31,6 +31,7 @@ void Host_NotifyMapLoaded();
 void Host_RefreshDSPDebuggerWindow();
 void Host_RequestRenderWindowSize(int width, int height);
 void Host_UpdateDisasmDialog();
+void Host_UpdateBreakpoints();
```
Please don't add anything new to the legacy Host interface. Instead, I'd suggest letting the UIs set a callback.

leoetlino commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -134,7 +143,7 @@ void GDBThread::OnStop()
     is_running = false;
 
-    gdb_signal(SIGTERM);
+    gdb_signal((u32)SignalTypes::Terminate);
```
`static_cast<u32>`

leoetlino commented:
Some of my comments show up as outdated, but they're still applicable to the branch.

sepalani commented on Source/Core/Core/PowerPC/TCPGecko.h:
```diff
@@ -0,0 +1,117 @@
+    static const u32 packetsize = 0xF800;
+    static const u32 uplpacketsize = 0xF80;
+
+    static const u8 cmd_poke08 = 0x01;
```
These seem only be used in a switch. Wouldn't it be better to be an enum (class)?

lioncash commented:
Personally I really don't think the extra 2500+ LOC increase is really worth supporting a (likely) niche usage case.

sepalani commented:
I have mixed feelings regarding this PR. On one hand, I really would like these features. On the other hand, I think the current Dolphin's debug interface is quite immature to fulfill that goal yet. So, I'd rather that to be fixed first before trying to implement that.

It seems your implementation is missing the StepOver() and StepOut() function. I don't recall MemoryBreakpoints to be working on Windows for me neither. What about symbol related features? Dolphin's also missing patching features, as well.

As said, it's quite niche and not very well documented nor stable due to Dolphin Debugger not being that great.

I also have some questions regarding those companion project:
- dolphindbg, I thought IDA already supported GDB interface, what does this plugin provide on top of it?
- geckowii, does it have anything to do with Chadderz's wiiu-tcp-gecko?

Anyway, feel free to correct me if I'm wrong.

leoetlino commented:
In my opinion, the improved gdbstub could be quite useful (since the current one is pretty much abandoned code which only works with the interpreter), but I think the TCP Gecko code should be part of another PR. This one is already big enough with just the GDB stub.

Also, please make it more obvious what was changed from the current gdbstub -- it looks like your new implementation is based on the old one but the diff isn't visible

revel8n commented:
Thanks for the feedback and suggestions.
Will work on getting those taken care of.

@sepalani 
Memory checks work fine for me in windows.
Stepping is usually handled on the remote side, not at the target, if I remember correctly from how gdb works.  But I could be wrong in this regard.
In that regard:
- StepInto and StepOver are handled on the remote side using breakpoints and information about the current instruction.
StepOut is not currently handled in any specific manner, so may not function as expected, as I do not make use of any stack frame or tracing to detect this case during execution.

my geckowii is not related to Chadderz and was originally modified before the wiiu was even released. Originally I did not have access to an actual USBGecko device to use with real hardware, and I had not started my plugins for ida pro in dolphin, so I wanted some way to debug and inspect running games.  So I modified the Gecko.Net project to use TCP instead of USB and connected to the interface within Dolphin.  The current TCPGecko implementation is based off of what I had originally just hacked into the EXI_DeviceGecko implementation in Dolphin, split into its own class.


--------

The main changes from the current GDBStub implementation are:
- fixed implementation of 'gdb_bp_ptr' to return correct array values (seems to be "broken" this way in most implementations, though I am not using it anyway (next bullet))
- Updated breakpoint handling to integrate directly with internal breakpoints and memory checks
- added additional handling of read, write, and access breakpoint signals
- extended the existing support for reading/writing the paired singles registers
- Updated execution state management to make use of CPU methods directly in order to work in both Interpreter and JIT
- added non-standard command for pausing execution

As for the rationale of the separation from the existing GDBStub:
Did not want to break the original functionality as some of my proposed changes make the interface/protocol non-standard (extra commands, registers, etc.).

The GDB remote serial protocol is convenient in that it is simple, the problem is that it is also usually "static", and often limited (usually limited to GPR and a few control registers, few implement vector, and other extended registers in the gdb stub implementations I have seen).  Depending on future needs would like to be able to add additional registers, etc. when it comes to remote emulation targets.

Both remote and target must know the layout of the data requested, and for the most part cannot be extended directly.

In this case not being able to access all the registers that would be useful limited what you could do without having to swap back to dolphin.  But changing the layout of the data means it is no longer compatible with other gdb implementation that may expect the data in some other way.


Most gdb stub implementations appear to be based off of the original floating around in some way or another, which is why mine is probably similar in that regard. When integrating mine I did update some things to match what was in the current GDBStub variant since it was more Dolphin specific than mine  (based off of work I have done with both the pcsx2 version, which I will be updating with the fixes from this one, and things found while working on the deci3dbg plugin for the ps3).

The TCPGecko implementation is based off of a combination of EXI_DeviceGecko.cpp in Dolphin and things done when converting Gecko.Net to support TCP.  So the suggestions made are interesting as I think some of those areas are directly from EXI_DeviceGecko.cpp, heh.

This was also originally made as a change back in the 4.0 branch just moved forward into the latest, so some things carried over that may have been cleanup during the transition.

In either case I will go through the suggestions more thoroughly and make the changes.

I would like to make it as useful as possible, but if I am in a minority in that regard I can continue to simply try and keep up with changes as they occur.

gyroninja commented:
Sorry for being ignorant, but isn't there already a tcp gecko implemented at Source/Core/Core/HW/EXI/EXI_DeviceGecko.cpp. Why do we need another one?

revel8n commented:
Still looking into how I want to handle the added "Host_" methods.

Hopefully I covered most the other feedback.

lioncash commented on Source/Core/Core/PowerPC/BreakPoints.h:
```diff
@@ -45,12 +47,21 @@ struct TWatch
 
+enum class MemCheckCondition
+{
+    MEMCHECK_NONE = 0x00,
```
This should adhere to the casing that @leoetlino used in his example [here](https://github.com/dolphin-emu/dolphin/pull/6325#discussion_r163789901). With a strongly-typed enum there's no need to repeat part of the name of the enum itself in its members (since the name itself already acts as a namespace).

revel8n replied:
Ah, holdover from pcsx2.  Been looking at it so long did not even notice.  Will fix.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+
+
+// private helpers
+static u8 hex2char(u8 hex)
```
These helpers should be adhering to our function naming convention.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+
+static void mem2hex(u8 *dst, u32 src, u32 len)
+{
+  u8 tmp;
```
This declaration can be joined with its assignment.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+
+static void hex2mem(u32 dst, u8 *src, u32 len)
+{
+  u8 tmp;
```
ditto

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+
+static void wbe32hex(u8 *p, u32 v)
+{
+  u32 i;
```
ditto

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+
+static void wbe64hex(u8 *p, u64 v)
+{
+  u32 i;
```
ditto

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+
+static u32 rbe32hex(u8 *p)
+{
+  u32 i;
```
ditto

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+
+static u64 rbe64hex(u8 *p)
+{
+  u32 i;
```
ditto

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+
+void GDBThread::gdb_read_command(void)
+{
+  u8 c;
```
This can be joined with its assignment

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+void GDBThread::gdb_read_command(void)
+{
+  u8 c;
+  u8 chk_read, chk_calc;
```
Ditto

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  }
+}
+
+void GDBThread::gdb_read_command(void)
```
`void` isn't necessary here in the parameter list.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  NOTICE_LOG(GDB_THREAD, "gdb: read command %c with a length of %d: %s\n", m_cmd_bfr[0], m_cmd_len, m_cmd_bfr);
+}
+
+void GDBThread::gdb_parse_command(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  }
+}
+
+u8 GDBThread::gdb_read_byte(void)
```
Ditto

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  return c;
+}
+
+u8 GDBThread::gdb_calc_chksum(void)
```
ditto

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  return c;
+}
+
+void GDBThread::gdb_reply(const char *reply)
```
asterisk should go against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+
+  u8 chk;
+  u32 left;
+  u8 *ptr;
```
asterisk against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  if (!m_connected)
+    return;
+
+  u8 chk;
```
Many of these can be joined with their assignments.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  if (!m_connected)
+    return 0;
+
+  size_t res;
```
This can be joined with its assignment.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+u8 GDBThread::gdb_calc_chksum(void)
+{
+  u32 len = m_cmd_len;
+  u8 *ptr = m_cmd_bfr;
```
Asterisk against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  strcpy(addr.sun_path, socket);
+
+  gdb_init_generic(PF_LOCAL, reinterpret_cast<const sockaddr *>(&addr), sizeof(addr),
+    NULL, NULL);
```
`nullptr`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+}
+
+void GDBThread::gdb_init_generic(int domain,
+  const sockaddr *server_addr, socklen_t server_addrlen,
```
asterisks against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+
+// GDB stub interface
+#ifndef _WIN32
+void GDBThread::gdb_init_local(const char *socket)
```
asterisks against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+    p[i] = nibble2hex(u8(v >> (28 - 4 * i)));
+}
+
+static void wbe64hex(u8 *p, u64 v)
```
Asterisk against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  }
+}
+
+static void wbe32hex(u8 *p, u32 v)
```
Asterisk against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  }
+}
+
+static void hex2mem(u32 dst, u8 *src, u32 len)
```
Asterisk against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+    return 'A' + n - 0xa;
+}
+
+static void mem2hex(u8 *dst, u32 src, u32 len)
```
Asterisk against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+
+#define REGISTER_ID(category, index) ((category << 8) | index)
+
+enum gdb_bp_type
```
This can likely be an `enum class`. This should also be cased as `GDBBPType`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  m_cmd_bfr[m_cmd_len + 2] = nibble2hex(chk >> 4);
+  m_cmd_bfr[m_cmd_len + 3] = nibble2hex(chk);
+
+  //DEBUG_LOG(GDB_THREAD, "gdb: reply (len: %d): %s\n", cmd_len, cmd_bfr);
```
This should be removed.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+
+  memset(m_cmd_bfr, 0, sizeof m_cmd_bfr);
+
+  m_cmd_len = (u32)strlen(reply);
```
`static_cast`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  }
+}
+
+void GDBThread::gdb_nak(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+    return;
+
+  const char nak = GDB_STUB_NAK;
+  size_t res;
```
Can be joined with its assignment.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+    fail("send failed");
+}
+
+void GDBThread::gdb_ack(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+    return;
+
+  const char ack = GDB_STUB_ACK;
+  size_t res;
```
Can be joined with its assignment.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  if (!m_connected)
+    return;
+
+  char bfr[128] = { 0 };
```
This can just be an empty brace assignment `= {};`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+    fail("send failed");
+}
+
+void GDBThread::gdb_handle_signal(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+
+  gdb_ack();
+
+  memset(bfr, 0, sizeof bfr);
```
This memset isn't necessary given its already zeroed beforehand.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  gdb_reply(bfr);
+}
+
+void GDBThread::gdb_continue(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  }
+}
+
+void GDBThread::gdb_detach(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  gdb_deinit();
+}
+
+void GDBThread::gdb_read_registers(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  if (!m_connected)
+    return;
+
+  u8 bfr[GDB_BFR_MAX - 4] = { 0 };
```
`= {};`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+    }
+  }
+
+  gdb_reply(reinterpret_cast<char *>(bfr));
```
asterisk against the type

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  gdb_reply(reinterpret_cast<char *>(bfr));
+}
+
+void GDBThread::gdb_write_registers(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  gdb_reply("OK");
+}
+
+void GDBThread::gdb_handle_set_thread(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+    memcmp(m_cmd_bfr, "Hc-1", 4) == 0 ||
+    memcmp(m_cmd_bfr, "Hc0", 4) == 0 ||
+    memcmp(m_cmd_bfr, "Hc1", 4) == 0)
+    return gdb_reply("OK");
```
This should probably be braced, given the condition travels more than one line.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  gdb_reply("E01");
+}
+
+void GDBThread::gdb_kill(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  NOTICE_LOG(GDB_THREAD, "killed by gdb");
+}
+
+void GDBThread::gdb_read_mem(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  if (!m_connected)
+    return;
+
+  u8 reply[GDB_BFR_MAX - 4] = { 0 };
```
`= {};`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+    return;
+
+  u8 reply[GDB_BFR_MAX - 4] = { 0 };
+  u32 addr, len;
```
These can be joined with their assignments.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+    gdb_reply("E01");
+
+  mem2hex(reply, addr, len);
+  gdb_reply(reinterpret_cast<char *>(reply));
```
asterisk against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  gdb_reply(reinterpret_cast<char *>(reply));
+}
+
+void GDBThread::gdb_write_mem(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  if (!m_connected)
+    return;
+
+  u32 addr, len;
```
These can be joined with their assignments.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  if (!m_connected)
+    return;
+
+  u8 reply[64] = { 0 };
```
`= {};`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  gdb_reply("OK");
+}
+
+void GDBThread::gdb_read_register(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+    }
+  }
+
+  gdb_reply(reinterpret_cast<char *>(reply));
```
asterisk against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  gdb_reply(reinterpret_cast<char *>(reply));
+}
+
+void GDBThread::gdb_write_register(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+    id = (id << 4) | hex2char(m_cmd_bfr[i++]);
+  ++i;
+
+  u8 * bufptr = m_cmd_bfr + i;
```
asterisk against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  gdb_reply("OK");
+}
+
+void GDBThread::gdb_handle_query(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  gdb_reply("");
+}
+
+void GDBThread::gdb_step(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  gdb_signal((u32)SignalTypes::Trap);
+}
+
+void GDBThread::gdb_add_bp(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  if (!m_connected)
+    return;
+
+  u32 type = 0, addr = 0, len = 0, i = 1;
```
These should be declared on their own lines. Some of these can also be joined with their first assignment.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  gdb_reply("OK");
+}
+
+void GDBThread::gdb_remove_bp(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  if (!m_connected)
+    return;
+
+  u32 type = 0, addr = 0, len = 0, i = 1;
```
These should be declared on their own line. Some of these can also be joined with their first assignment.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1322 @@
+  gdb_reply("OK");
+}
+
+void GDBThread::gdb_pause(void)
```
Ditto about `void` not being needed here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,131 @@
+#endif
+#include <stdarg.h>
+
+#define		GDB_BFR_MAX	10000
```
This can likely be a private static constexpr class variable

lioncash commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,131 @@
+
+#pragma once
+
+#include <signal.h>
```
Many of these includes are likely not necessary in the header.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,131 @@
+  void ThreadFunc();
+
+protected:
+  bool m_is_running;
```
Class variables should follow after functions

lioncash commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,131 @@
+  CPU::State m_previous_state;
+
+  bool m_connected;
+  int m_tmpsock = -1;
```
Does it make sense for only these two to have default initial values?

lioncash commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,131 @@
+
+  void gdb_init(u32 port);
+#ifndef _WIN32
+  void gdb_init_local(const char *socket);
```
asterisk against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,131 @@
+  u64 m_last_signal_address;
+  MemCheckCondition m_last_signal_condition;
+
+  void gdb_init(u32 port);
```
These should follow our function casing style.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,131 @@
+
+  void ThreadFunc();
+
+protected:
```
This `protected` isn't necessary

lioncash commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,131 @@
+  bool Initialize();
+  void Terminate();
+
+protected:
```
Does it make sense for this to be protected rather than private? If the answer is yes, then the destructor should be virtual.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,131 @@
+#endif
+  void gdb_deinit();
+
+  bool gdb_active();
```
This can be a const member function.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,131 @@
+  u8 gdb_read_byte();
+  u8 gdb_calc_chksum();
+
+  void gdb_reply(const char *reply);
```
asterisk against the type.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,131 @@
+  void gdb_signal(u32 s, u64 addr = -1, MemCheckCondition cond = MemCheckCondition::MEMCHECK_NONE);
+
+  void gdb_init_generic(int domain,
+    const sockaddr *server_addr, socklen_t server_addrlen,
```
asterisk against the type.

revel8n commented:
@lioncash 
Hmm. Ok.  With the suggestions related to the gdb helpers, I guess I will ask: Is the current GDBStub used in any way?

I had originally included these methods here solely to not conflict with the original GDBStub while I validated my implementation, as I did not know whether the original was actually used by anyone in its current form and did not want to break its functionality.

If not used currently, or desired for my changes to replace it anyway, would it be ok then that I simply refactor the existing GDBStub with the changes you are suggesting, and then just call that from GDBThread?

Since the functions mentioned are all from that code originally, a lot of it is simply holdover from there.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1296 @@
+#define		GDB_STUB_ACK	'+'
+#define		GDB_STUB_NAK	'-'
+
+enum class SignalTypes
```
This should be singular `SignalType`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1296 @@
+  }
+}
+
+void GDBThread::GDBSignal(u32 s, u64 addr, MemCheckCondition cond)
```
`s` and `m_last_signal` should likely be a `SignalType`. This would remove the need for casting at every callsite.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1296 @@
+    PowerPC::breakpoints.Remove(addr);
+  }
+
+  INFO_LOG(GDB_THREAD, "gdb: removed a %d breakpoint: %08x bytes at %08X\n", static_cast<std::underlying_type_t<GDBBPType>>(type), len, addr);
```
You can just cast to int here.

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1296 @@
+  m_sock = -1;
+  m_cmd_len = 0;
+  m_last_signal = 0;
+  m_last_signal_address = -1;
```
`= 0xFFFFFFFFFFFFFFFF` or `UINT64_MAX`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1296 @@
+  }
+  else if (std::get<0>(triggered_breakpoint))
+  {
+    u64 addr = std::get<1>(triggered_breakpoint).value_or(-1);
```
`0xFFFFFFFFFFFFFFFF` or `UINT64_MAX` instead of `-1`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1296 @@
+  return res;
+}
+
+static u64 ReadHexBE64(u8* p)
```
`p` can be a `const u8*`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1296 @@
+
+  GDBAck();
+
+  u8* bufptr = m_cmd_bfr + 1;
```
`bufptr` can be a `const u8*`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1296 @@
+    id = (id << 4) | HexToChar(m_cmd_bfr[i++]);
+  ++i;
+
+  u8* bufptr = m_cmd_bfr + i;
```
`bufptr` can be a `const u8*`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1296 @@
+    type = (type << 4) | HexToChar(m_cmd_bfr[i++]);
+  i++;
+
+  GDBBPType breakpointType = GDBBPType::None;
```
`breakpoint_type`

lioncash commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1296 @@
+{
+  bool is_mem_check = false;
+
+  TMemCheck MemCheck;
```
`mem_check`

filfat commented:
Would be amazing to see the work here continued 👍

BhaaLseN commented on Source/Core/Core/Core.cpp:
```diff
@@ -346,11 +347,22 @@ static void CpuThread(const std::optional<std::string>& savestate_path, bool del
 
   s_is_started = false;
 
+  if (_CoreParameter.bEnableDebugging)
```
Thoughts on checking whether `gdb_thread` is valid/alive instead of checking the parameter? Not quite sure what happens if (or when) that value changes after the CPU is started.

revel8n replied:
True.  Some of the logic will need to be updated to take that possibility into account.  Since i had always been using the command line parameter which turns this on, did not remember about the fact it was exposed from within the UI as well. Will take a look and update accordingly.

BhaaLseN commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1274 @@
+    return;                                                                                        \
+  }
+
+#define failr(msg)                                                                                 \
```
Any thoughts on making those into methods? Using defines for this purpose is kinda last century :D

revel8n replied:
Hold over from plugin based on older C based code sample where implementation was first made and this was merged from.  Believe i fixed this in all other versions of this except for the one for Dolphin, heh.

BhaaLseN commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1274 @@
+#define GDB_STUB_START '$'
+#define GDB_STUB_END '#'
+#define GDB_STUB_ACK '+'
+#define GDB_STUB_NAK '-'
```
Those might make more sense as `constexpr`.

revel8n replied:
Inherited from Dolphin's GDBStub implementation, which probably inherited it from the original C based gdbstub implementation. Will take a look.

BhaaLseN commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1274 @@
+#endif
+#endif
+
+#define REGISTER_ID(category, index) ((category << 8) | index)
```
I don't see this one being used?

revel8n replied:
Inherited from plugin code where client implementation resides.  Will remove.

BhaaLseN commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1274 @@
+
+void GDBThread::UpdateState(CPU::State current_state)
+{
+  if (current_state != m_previous_state)
```
You could invert this check and exit early, to get rid of a level of indentation:
```suggestion
  if (current_state == m_previous_state)
    return;
```

revel8n replied:
True. Will make the change. Thanks.

BhaaLseN commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1274 @@
+
+  m_tmpsock = socket(domain, SOCK_STREAM, 0);
+  if (m_tmpsock == -1)
+    ERROR_LOG(GDB_THREAD, "Failed to create gdb socket");
```
Should you return here when creating the socket fails?
And...kinda same question for all the following ones (which will likely fail as well if you don't return on the previously failed ones)

revel8n replied:
Inherited from Dolphin's GDBStub implementation, which probably inherited it from the original C based gdbstub implementation. Will take a look.

BhaaLseN commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1274 @@
+  int on = 1;
+  if (setsockopt(m_tmpsock, SOL_SOCKET, SO_REUSEADDR, reinterpret_cast<const char*>(&on),
+                 sizeof on) < 0)
+    ERROR_LOG(GDB_THREAD, "Failed to setsockopt");
```
In case this one sticks around as a one-liner, it should still be braced (since the condition spans over two lines)

revel8n replied:
Inherited from Dolphin's GDBStub implementation. Was originally one line until previous comment suggested updating to reinterpret_cast, which caused auto formatting to push it to second line. Will take a look.

BhaaLseN commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1274 @@
+  if (m_tmpsock != -1)
+  {
+    shutdown(m_tmpsock, SD_BOTH);
+#if _WIN32
```
This if/else around `closesocket`/`close` appears fairly often, so it might be worthwhile wrapping it in a method to reduce the `#ifdef` cruft in other places (which makes it somewhat difficult to follow the intent of the code IMO; but that might just be me)

revel8n replied:
Overall method inherited from Dolphin's GDBStub implementation, but updated to fix compile issues. Makes sense to limit duplication of the defines and logic.

BhaaLseN commented on Source/Core/Core/PowerPC/GDBThread.h:
```diff
@@ -0,0 +1,148 @@
+  int m_tmpsock = -1;
+  int m_sock = -1;
+
+  u8 m_cmd_bfr[GDB_BFR_MAX + 1];
```
Thoughts on using `std::array` instead?

revel8n replied:
Inherited from Dolphin's GDBStub implementation, which probably inherited it from the original C based gdbstub implementation. Had avoided making too many changes to its implementation, but i suppose there is no reason not to update things.

BhaaLseN commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1274 @@
+
+  m_cmd_len++;
+  u8 chk = GDBCalculateChecksum();
+  m_cmd_len--;
```
Just wondering why this place generates a checksum for the current command buffer size **plus one** but the other places don't.
It would probably be clearer if you straight up updated `m_cmd_len` after the `memcpy` (which does indeed skip the first byte and and start writing at the 2nd, assuming because it is some sort of command indicator) and continue with it for the remainder of the method.

But then again, you're also writing to `m_cmd_len + 3` (or +2 if you drop this decrease again) _without_ updating the length at all, so this is somewhat confusing (at least to me).

This probably _at least_ deserves a comment or two, for anyone that comes afterwards.

revel8n replied:
Inherited from Dolphin's GDBStub implementation, which probably inherited it from the original C based gdbstub implementation. Will take a look.

BhaaLseN commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1274 @@
+  if (!m_connected)
+    return;
+
+  char bfr[128] = {};
```
No need to hold back on letters ;)
```suggestion
  std::array<char, 128> buffer = {};
```

revel8n replied:
Inherited from Dolphin's GDBStub implementation, which probably inherited it from the original C based gdbstub implementation. Had avoided making too many changes to its implementation, but i suppose there is no reason not to update things.

BhaaLseN commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1274 @@
+    break;
+  default:
+    return;
+    break;
```
You might want to remove this `break`, since some compiler might mark it as unreachable code after that `return`

revel8n replied:
Makes sense.

BhaaLseN commented on Source/Core/Core/PowerPC/GDBThread.cpp:
```diff
@@ -0,0 +1,1274 @@
+
+  u8* bufptr = bfr;
+
+  for (u32 id = 0; id <= 71; ++id)
```
There's a lot of magic numbers down here; would any of them benefit from being an `enum class` or at least some named `constexpr`? Having a name or something else to hold on to might make it easier to read/understand the code later on

revel8n replied:
Overall method inherited from Dolphin's GDBStub implementation. Magic numbers duplicated when logic was duplicated from the original Read/Write Register (singular) methods to the Read/Write Registers (plural) methods in order for them to support the same registers.  Makes sense to limit this going forward.  Will see if any of Dolphin's exiting enumerations or defines can help here and/or add as needed.

Tilka commented:
ping

dreamsyntax commented:
Are we good to close this since #10142 may have made it obsolete?
https://github.com/dolphin-emu/dolphin/pull/10142#issuecomment-948193166

dreamsyntax commented:
Are we good to close this since #10142 may have made it obsolete?
https://github.com/dolphin-emu/dolphin/pull/10142#issuecomment-948193166

@phire

AdmiralCurtiss commented:
Closing on assumption of being obsolete.
